### ISSUE TITLE
Long execution Time

### ISSUE THREAD

**Issue by aschi75** [vmware/webcommander#62]
The Execution time of Scripts is very long.
I used the following Script for testing with a simple output “END”

Param (
	$username,
	$instance="MMA-Saturn"
)
foreach ($paramKey in $psboundparameters.keys) {
	$oldValue = $psboundparameters.item($paramKey)
	$newValue = [system.web.httputility]::urldecode("$oldValue")
	set-variable -name $paramKey -value $newValue
}
. .\objects.ps1
writeCustomizedMsg "END" 

Result:
![result](https://cloud.githubusercontent.com/assets/12539205/8230828/1c22e926-15c4-11e5-82a2-0efbd16daada.png)

I did some more Test, I just run the Script in Webcommander again. During the Tests I logged in/out from the Webserver with the User who started the scripts with webcommander:

Test 2: 128.9 Seconds
Test 3: 128.8 Seconds
Test 4: User added to local Administrator Group on Webserver -> 120.5 s
Test 5: User logged in -> 120.6s
Test 6: User logged in, tested Internet access, started a powershell Sessoin -> 11.2 s
Test 7: User still logged in but closed powershell Session -> 3.5s
Test 8: User logged out -> 108.8s
Test 9: Adminstrator logged in and started a Powershell Session -> 109.2s
Test 10: User logged in -> 4.1s
Test 11: User logged out -> 108.9s
Test 12: User logged in -> 3.6s
Test 13: User logged out -> 109.1s 
Test 14: deleted User Profile on the Webserver, User logged in afterwards -> 108.6s
Test 15: logged in User started a powershell Sessoin -> 108.8s
Test 16: Logged in User started www.google.ch -> Popup IE Enhanced Security Message -> 4.6s
Test 17: Logged out deleted User Profile again. 108.8s
Test 18: Logged in and turned Enhanced IE Security OFF -> 108.6s
Test 19: started www.google.ch, no Popup window appeared -> 3.9s

The execution time is good if the following conditions are met:
A User who is starting a Script is logged in to the Webserver AND had once started IE. the user must be logged in on the Webserver. Without these conditions it takes almost more than 2 Minutes for the Scripts to execute.
But Users are not allowed to login to the Webserver. And because they need to be logged in, it’s even worse.
Did someone else face this Problem? What Solutions are possible?

**Comment by 9whirls** [comment]
Hi Roger,
How long does it take to run the script from powershell console?
> cd c:\webcommander\powershell
> .\yourscript.ps1

**Comment by aschi75** [comment]
Hi deleted the User Profile on the Webserver again. The execution time with Webcommander is over 100 Seconds.
The Powershell console started fom a command prompt is immediately available. I started it with the User who is also using Webcommander.

I used the Measure-Command to measure the execution time within a Powershell Session on the Webserver.
![measure](https://cloud.githubusercontent.com/assets/12539205/8248066/dce97850-165a-11e5-8189-f2401df40534.PNG)
If I start the script from a powershell session and not from "Powershell ISE" it's even a little faster.

I Started IE and the Execution time afterwards with Webcommander was 4.2 Seconds. 
The first execution in Powershell is always about 300Milliseconds and goes down to about 40 Milliseconds if I restart the same command. It does not matter if IE was ever started or not.

**Comment by jrob24** [comment]
What version of IE do you have installed on WebCommander server? Also what do you have in the global located under C:\Windows\SysWOW64\WindowsPowerShell\v1.0?

**Comment by 9whirls** [comment]
So there seems nothing time consuming in your powershell environment.
Could you try install firefox or chrome on the server and run the command there?
Make sure no proxy is configured in the browser.

I know you have multiple webcommander servers in your lab. 
Do you hit the long execution time issue on all of them?

**Comment by aschi75** [comment]
Hi I have IE 11 installed:
![ie](https://cloud.githubusercontent.com/assets/12539205/8277059/954911d4-18be-11e5-858e-ec639ff187fc.PNG)

I don't have a File or Folder called global located in the Path you specified. What did you mean?
![global](https://cloud.githubusercontent.com/assets/12539205/8277107/569e83dc-18bf-11e5-9001-8a13fd494bcd.png)

The execution time on another Server is "just 85 Seconds". I don't know why it's 20 Seconds faster. If I logged in to the Webserver with the user and started once IE. The execution Time is also 3.8 Seconds.

I installed Firefox and deleted the User Profile. The startup of the webcommander takes very long because it tries to connect to different WebPages which is not possible with no Proxy Settings (jschr.github.io...., font.googleapis.com..., cdn.datatables.net..., code.jquery.com. The "RUN Button" is not active, probably because of not available java scripts. Is There a Webcommander Package with no need to be able to connect to the Internet? Maybe the reason for the long execution time has someting to do with this.

**Comment by 9whirls** [comment]
Actually, your webcommander server doesn't have to have access to internet. Your client machine does. You could execute commands from another machine who has internet access. 

If that is not available, your could put all those internet files (.js or .css) locally on your webcommander server. If you do so, make sure to edit .html and .xsl files to change the reference path to them.

Please also note that you don't even have to use browser to execute command on webcommander. Those js and css files only are meaningful to browser. If you trigger commands via program, such as CURL or invoke-webrequest, those files are ignored.

**Comment by aschi75** [comment]
Ok, I just thought because you said I should test webcommander directly on the webserver. Clients have Internet Access through a Proxy. The Proxy doesn't block anything. I tracked it also with developer Tools and Fiddler but couldn't find any requests.
![f12](https://cloud.githubusercontent.com/assets/12539205/8299432/9d51ce74-1978-11e5-9676-fe85bd038016.PNG)
![fiddler](https://cloud.githubusercontent.com/assets/12539205/8299434/9fde5428-1978-11e5-8684-75da68325e83.PNG)

I checked some URLs from webCmd.xsl like http://cdn.datatables.net/1.10.4/css/jquery.dataTables.min.css or http://code.jquery.com/ui/1.9.2/themes/base/jquery-ui.css which are accessible from the Client.

I enabled anonymous Authentication, changed the NTFS Permissions accordingly and started the webrequest with curl with the same result.
![curl](https://cloud.githubusercontent.com/assets/12539205/8299602/470a9714-197b-11e5-9f54-72f2f40cf8ca.PNG)

**Comment by 9whirls** [comment]
The direct test is prefered because no proxy is involved. However, your server has no internet access so we need to use another client machine.

The execution time from webcommander perspective is basically the time it takes to run the powershell script. It should roughly equal to the time you run the script from powershell console + the console startup time. I higlight the console startup time because some people have special powershell profile which may take long time to run everytime they start a powershell console.

However, based on your observation, it seems you could start powershell console and run the script both very quick. So I have no clue why it takes as long as 65 seconds to run a simple command. On my server, it takes just 2 seconds.

Is there any other process on the server consuming large CPU time? If not, I may have to have a look at your server to make further investigation.

**Comment by jrob24** [comment]
Sorry it looks like I found a to include the word profile in my last response. I was asking about what do you have inside the global profile under SYSWOW64\WindowsPowerShell\v1.0?

You may want to test execution with adding the following to the global profile and see if its speeds anything up.

`Set-ItemProperty -path "HKCU:\Software\Microsoft\Windows\CurrentVersion\WinTrust\Trust Providers\Software Publishing" -name State -value 146944`

http://www.virtu-al.net/2010/07/05/speed-up-powercli/

**Comment by aschi75** [comment]
Hi, sorry the delay. There are no other process on the Server consuming large CPU. the Server has 4 CPUs and 6GB Memory. The Webserver itself is almost empty and idle.Yes, Script runs quick in the powershell console.

I added the mentionend Line in C:\Windows\SysWOW64\WindowsPowerShell\v1.0\Profile.ps1 but the execution time is still about 60 Seconds.
BUT: once I did the mentioned IE Settings manually with the logged in User it worked even when I logged out afterwards. I checked if I have the same value in the Registry Key when I disable this option -> Yes. I deleted the Profile again to test the whole process again. First i started the script which should take again about 60Seconds but i finished within 4 Seconds. Probably the change I made in the Profile.ps1 wasn't active immediately.
For now it looks good. I checked also the other Script. Perfect. Thanks lot for  your Support, I'm happy
